src/main/java/org/jvnet/hudson/test/InboundAgentRule.java

```diff
@@ -0,0 +1,111 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jvnet.hudson.test;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import hudson.model.Slave;
+import hudson.remoting.Launcher;
+import hudson.remoting.Which;
+import hudson.slaves.DumbSlave;
+import hudson.slaves.JNLPLauncher;
+import hudson.slaves.RetentionStrategy;
+import hudson.util.StreamCopyThread;
+import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.tools.ant.util.JavaEnvUtils;
+import org.junit.rules.ExternalResource;
+
+/**
+ * Manages inbound agents.
+ * While these run on the local host, they are launched outside of Jenkins.
+ * @see JenkinsRule#createComputerLauncher
+ * @see JenkinsRule#createSlave()
+ */
+public final class InboundAgentRule extends ExternalResource {
+
+    private final Map<String, Process> procs = new HashMap<>();
+
+    /**
+     * Creates, attaches, and starts a new inbound agent.
+     * @param name an optional {@link Slave#getNodeName}
+     */
+    @SuppressFBWarnings(value = "PATH_TRAVERSAL_IN", justification = "just for test code")
+    public Slave createAgent(@NonNull JenkinsRule r, @CheckForNull String name) throws Exception {
+        if (name == null) {
+            name = "agent" + r.jenkins.getNodes().size();
+        }
+        // TODO 2.216+ support WebSocket option
+        DumbSlave s = new DumbSlave(name, new File(r.jenkins.getRootDir(), "agent-work-dirs/" + name).getAbsolutePath(), new JNLPLauncher(true));
+        s.setNumExecutors(1); // TODO pending 2.234+
+        s.setRetentionStrategy(RetentionStrategy.NOOP);
+        r.jenkins.addNode(s);
+        start(r, name);
+        return s;
+    }
+
+    /**
+     * (Re-)starts an existing inbound agent.
+     */
+    @SuppressFBWarnings(value = "COMMAND_INJECTION", justification = "just for test code")
+    public void start(@NonNull JenkinsRule r, String name) throws Exception {
+        stop(name);
+        ProcessBuilder pb = new ProcessBuilder(JavaEnvUtils.getJreExecutable("java"), "-Djava.awt.headless=true", "-jar", Which.jarFile(Launcher.class).getAbsolutePath(), "-jnlpUrl", r.getURL() + "computer/" + name + "/slave-agent.jnlp");
```

Comment:
This uses JVM ergonomics to set the maximum heap size and provides no observability. In the past, these deficiencies have led to the development of #380 and #425. If we are introducing a new API and expecting consumers to use it, I think it should not come with these deficiencies which have been already identified and solved elsewhere.

Reply:
I will take a look at those and try to make matching changes. Note that this code is taken verbatim from `workflow-durable-task-step-plugin` where it has been running without apparent issues for 7½y (https://github.com/jenkinsci/pipeline-plugin/commit/cd1572249248608ef9ccb904c99c707ee8e0a655).